synapse/handlers/device.py

```diff
@@ -367,6 +381,22 @@ async def check_device_registered(
 
         raise errors.StoreError(500, "Couldn't generate a device ID.")
 
+    async def _delete_stale_devices(self) -> None:
+        """Background task that deletes devices which haven't been accessed for more than
+        a configured time period.
+        """
+        # We should only be running this job if the config option is defined.
+        assert self._delete_stale_devices_after is not None
+        now_ms = self.clock.time_msec()
+        since_ms = now_ms - self._delete_stale_devices_after
+        devices = await self.store.get_devices_not_accessed_since(since_ms)
+
+        for user_id, user_devices in devices.items():
+            # The devices table can contain devices for remote users, and we don't want
+            # e.g. break encryption by accidentally removing them.
```

Comment:
From a glance it looks like this is only the case for hidden devices, but I'd rather be overly safe here.

Reply:
This comment seems to say that we *shouldn't* have hidden devices for remote users?

https://github.com/matrix-org/synapse/blob/2cc5ea933dbe65445e3711bb3f05022b007029ea/synapse/storage/databases/main/end_to_end_keys.py#L1236-L1241

Reply:
I see this now, but interestingly my homeserver's database seems to think otherwise:

```
synapse=# select count(*) from devices where hidden is TRUE and user_id like '%abolivier.bzh';
 count 
-------
    30
(1 row)

synapse=# select count(*) from devices where hidden is TRUE and user_id not like '%abolivier.bzh';
 count 
-------
   813
(1 row)
```

Reply:
Maybe due to #6956? That talks a bit about adding a background update for this, but chose not to?

Reply:
Actually, would we want this check in `get_devices_not_accessed_since` instead of here? I guess it doesn't make too much of a difference, but I don't think remote devices are ever "stale"?

Reply:
Yes that's a good point.

Reply:
> Maybe due to #6956? That talks a bit about adding a background update for this, but chose not to?

Right, yes, that looks like it's the culprit.

Reply:
I filed #12855 for this.